publish-*artifacts: merge implementations of putRelease / putNonRelease

otan opened the pull request:
The `putRelease` / `putNonRelease` functionality of `publish-artifacts`
and `publish-provisional-artifacts` are almost exactly the same. I've
merged this to one implementation in the `release` package. This change
should have no functional difference to what is being published /
archived today.

I've also added the ability to put multiple files into a release archive
/ to be sent as part of a non release. This is relevant for a future
change, but getting that work bootstrapped here.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50900)
<!-- Reviewable:end -->

otan commented:
Publish Bleeding Edge: https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_MergeToMaster&branch_Cockroach=50900&tab=buildTypeStatusDiv

otan committed:
publish-*artifacts: merge implementations of putRelease / putNonRelease

The `putRelease` / `putNonRelease` functionality of `publish-artifacts`
and `publish-provisional-artifacts` are almost exactly the same. I've
merged this to one implementation in the `release` package. This change
should have no functional difference to what is being published /
archived today.

I've also added the ability to put multiple files into a release archive
/ to be sent as part of a non release. This is relevant for a future
change, but getting that work bootstrapped here.

Release note: None

jlinder commented on pkg/release/release.go:
```diff
@@ -146,3 +160,256 @@ func MakeRelease(b SupportedTarget, pkgDir string, opts ...MakeReleaseOption) er
+	PutObject(*s3.PutObjectInput) (*s3.PutObjectOutput, error)
+}
+
+// S3KeyRelease is extracts the target archive base and archive
```
```suggestion
// S3KeyRelease extracts the target archive base and archive
```

otan replied:
That's a neat GH feature!

jlinder commented on pkg/cmd/publish-provisional-artifacts/main.go:
```diff
@@ -58,23 +54,18 @@ func makeS3() (s3I, error) {
 var doBlessF = flag.Bool("bless", false, "bless provisional binaries")
 
 var (
-	noCache = "no-cache"
```
Why only move `noCache` from here and not from `publish-artifacts`?

otan replied:
Oops! Done!

jlinder reviewed (approved):
LGTM

Thanks for consolidating!

otan commented:
TFTR!

bors r=jlinder

otan commented:
bors r=jlinder
